### PR TITLE
Support '*' in CORS configuration

### LINKED ISSUE #5422: quarkus.http.cors.origins=* should be supported
**Describe the bug**
When I read the documentation about configuring cors for the first time, I tought that leaving "quarkus.http.cors.origins" as empty would be the same as "Access-Control-Allow-Origin: *".
The * wildcard allows any origin to access ressources.

Furthermore, a value of "*" is currently not supported. See CORSFilter.java#L93.
`boolean allowsOrigin = corsConfig.origins.isEmpty() || corsConfig.origins.contains(origin);`

We should try to improve the documentation to reflect what is actually going on in the filter.

**Expected behavior**

> For requests without credentials, the literal value "*" can be specified, as a wildcard; the value tells browsers to allow requesting code from any origin to access the resource.

If I use "*" as wildcard, I expect quarkus to allow any origin to access my resources.
If I remove "quarkus.http.cors.origins" in my configuration, I expect quarkus to use "*" as default.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin

**Actual behavior**
Requests get rejected, altough quarkus.http.cors.origins is set to "*".

**Configuration**
```properties
uarkus.http.cors=true
quarkus.http.cors.origins=*

```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
SYS_NT-10.0 DESKTOP-0TKUM9I 2.11.2(0.329/5/3) 2018-11-10 14:38 x86_64 Msys
 - Output of `java -version`:
java version "1.8.0_211"
Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.211-b12, mixed mode)
 - Quarkus version or git rev: 1.0.0.CR1

### PR DESCRIPTION
**Pull request by xumk**
Resolve #5422 
I update change from [request](https://github.com/quarkusio/quarkus/pull/5455) @Postremus for new quarkus version and fix remarks

### CODE REVIEW

**Comment by sberyozkin** [comment]
@xumk Can you please squash the commits ?

**Review by sberyozkin — APPROVED** [review]
LGTM, but would appreciate Andy @ia3andy checking it out as well

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:3]: We avoid star imports.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:129]:
```suggestion
        return list.isEmpty() || (list.size() == 1 && "*".equals(list.get(0)));
```
always better to be explicit.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java:33]: Why this change, it doesn't look related to dealing with `*`?

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks! Needs a squash/rebase but I also spotted some issues in the PR.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (xumk) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java:33]: @gsmet it change was same added possibility  use * for cors methods. Can you promt me how best to do it?

**Comment by xumk** [comment]
> @xumk Can you please squash the commits ?
Can you help me to do it please? I am beginner git user :)

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (ia3andy) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java:33]: I think it's because `*` is not in the `HttpMethod` Enum

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (ia3andy) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: Yeah I think this is important too

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (ia3andy) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:62]: Maybe you can still convert them to `HttpMethod` Enum to avoid having invalid inputs?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (ia3andy) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:127]: do you have a better label than `containsOnlyWildcard` for this method?

maybe something like `isConfiguredWithWildcard`

**Comment by sberyozkin** [comment]
@xumk Hi, sure, this is how I do it, Guillaume who is the real guru may have a much better recommendation, so:

1. rebase
```
git fetch upstream master
git rebase upstream/master
# do it in case of a conflict, open an editor, fix it, and then
# git add path/to/OidcTenantConfig.java
# git rebase --continue
git push origin your_branch --force
```
2. squash
```
git log --abbrev-commit
git rebase -i sha-of-commit-immediately-before-your-first-commit-on-your-branch
// this will open a vi editor, replace all `pick` command except the very first `pick your_first_commit` with `f`, and save - this will complete the squash
git push origin your_branch --force
```

**Review by ia3andy — CHANGES_REQUESTED** [review]
I added a few suggestions...

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (ia3andy) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:127]: Maybe you could make this static and add a small test making sure it responds correctly?

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:127]: > Maybe you could make this static and add a small test making sure it responds correctly?

Actually, it is a good one. This should have been tested indirectly in this `CORSWildcardSecurityTestCase` but now that I've seen this suggestion I've realized `CORSWildcardSecurityTestCase` needs its application properties updated. 
@xumk please update it to check both wildcard origins and headers

**Comment by sberyozkin** [comment]
@xumk While you on this Cors PR, can you please update it to trim all the values which have been produced by `COMMA_SEPARATED_SPLIT_REGEX` ? I can do it easily too, but then we'll need to resolve the conflicts :-)

**Comment by xumk** [comment]
@sberyozkin i did squash, but i don't know how to do squash merge :( can you see I understood correctly add call trim method for strings produced by COMMA_SEPARATED_SPLIT_REGEX?

**Comment by sberyozkin** [comment]
@xumk @stuartwdouglas has opened a different PR, #12177. So please remove your `trim()` code, I don't really mind if it stays, but it will be redundant once #12177 is merged. Sorry I was not aware of #12177 
@gsmet There is a strange JDK 11 error, let me rerun

**Comment by sberyozkin** [comment]
The same issue is here, #12177 so not related to this one

**Comment by sberyozkin** [comment]
@xumk Can you please add one more test, say `CORSWildcardStarSecurityTestCase`, `CORSWildcardSecurityTestCase` does not test the properties containing `*`

**Review by ia3andy — APPROVED** [review]
LGTM (needs squash)

**Comment by xumk** [comment]
@sberyozkin things are good. thank you for help and remarks

File: extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSWildcardStarSecurityTestCase.java
Comment (ia3andy) [inline extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSWildcardStarSecurityTestCase.java:21]: I guess you could extend `CORSWildcardSecurityTestCase.java` and just overrides the config?

Comment (xumk) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSWildcardStarSecurityTestCase.java:21]: @ia3andy I thought which will be better it have two test for two configuration. First test configuration is not include this properties and second test configuration is include this properties with * values

Comment (ia3andy) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSWildcardStarSecurityTestCase.java:21]: ah ok, I misread it ;-)

**Comment by sberyozkin** [comment]
@xumk great, I see Andy has approved so if Guillaume @gsmet is also happy then we can merge.
Not sure about `trim()` against the request header values, I'd assume Vert.x should take care of it.

You probably need to pick the JDK 11 build fix from Guillaume, so please rebase again and also update the commit message, small typo is there, from `add support start for cors filter` to something similar to your PR subject, maybe `Support '*' in CORS filter configuration`, something like that :-)

So rebase and then `git commit --amend -m "The updated message"`, and force-push

thanks

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (gsmet) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java:33]: Ah OK, totally missed it, sorry.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
I rebased and force pushed.

**Comment by jaikiran** [comment]
Removed the "waiting-for-ci" since this now needs a rebase against master due to a merge conflict.

**Comment by gsmet** [comment]
I rebased and fixed the conflict.

**Pull request merged**
